### PR TITLE
feat: add table component

### LINKED ISSUE #48: Table Component
## Expected Behavior
Make a `<Table>` Component
Features:
- Should come with default `table` props
- Inherit from `<table>`

## Current Behavior
No Table component

- [ ] Make spec for Table component
- [ ] Make the base component with API and without styles
- [ ] add styling

### PR DESCRIPTION
**Pull request by hafsahayath**
## Description
- Adds table component to the library

## How Has This Been Tested?
- tested it in the examples folder.

## Screenshots (if appropriate):
- basic table
image

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

closes #48

### CODE REVIEW

File: lib/src/components/Table/Table.css.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import { globalStyle, style, createTheme } from "@vanilla-extract/css";
+import { ITableContainerThemeVars, ITableThemeVars } from "./Table.types";
+
+export const [, tableContainerThemeVars]: ITableContainerThemeVars =
+  createTheme({
+    customBorderRadius: "4px",
+    customBorderColor: "black",
+  });
+
+export const [, tableThemeVars]: ITableThemeVars = createTheme({
```
Comment (pruthvi2103) [inline lib/src/components/Table/Table.css.ts:10]: descructure the class name and add it to the container. Helps scope the variables

File: lib/src/components/Table/Table.css.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  textAlign: tableThemeVars.headerAlignment as "left" | "right" | "center",
+});
+
+globalStyle(`${tableDefaults} td`, {
+  textAlign: tableThemeVars.contentAlignment as "left" | "right" | "center",
+});
+
+globalStyle(`${tableDefaults}.striped tr:nth-child(even)`, {
+  backgroundColor: tableThemeVars.customStripeColor,
+});
```
Comment (pruthvi2103) [inline lib/src/components/Table/Table.css.ts:48]: Seperate global styles in a file called `.global.styles.css.ts`

File: lib/src/components/Table/Table.tsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    [tableThemeVars.contentAlignment]: alignContent,
+    [tableThemeVars.headerAlignment]: alignHeading,
+    [tableThemeVars.customCellPadding]: cellPadding,
+  });
+
+  return (
+    <div
+      style={{ ...tableContainerInlineVars }}
+      className={`${tableContainerDefaults} ${overrideTableContainerClass}`}
+    >
```
Comment (pruthvi2103) [inline lib/src/components/Table/Table.tsx:48]: any reason why the table has to be wrapped in a `div` and not directly a `table` ?

File: lib/src/components/Table/TableCaption/TableCaption.tsx [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import { TableCaptionProps } from "./TableCaption.types";
+
+const TableCaption: ForwardRefRenderFunction<
+  HTMLTableCaptionElement,
+  TableCaptionProps
+> = ({ children, className, placement = "bottom", ...nativeProps }, ref) => {
+  return (
+    <caption
+      ref={ref}
+      style={{ captionSide: placement, paddingTop: "10px" }}
```
Comment (pruthvi2103) [inline lib/src/components/Table/TableCaption/TableCaption.tsx:11]: native props will overwrite the style here, merge the prop style and this.

File: lib/src/components/Table/Table.tsx [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    [tableThemeVars.customRowHeight]: rowHeight,
+    [tableThemeVars.contentAlignment]: alignContent,
+    [tableThemeVars.headerAlignment]: alignHeading,
+    [tableThemeVars.customCellPadding]: cellPadding,
+  });
+
+  return (
+    <div
+      style={{ ...tableContainerInlineVars }}
+      className={`${tableContainerDefaults} ${overrideTableContainerClass}`}
```
Comment (pruthvi2103) [inline lib/src/components/Table/Table.tsx:47]: classname will be overWritten by native props is user gives it, destructure it and add the prop className and add it at end (refer something like avatar)

File: lib/src/components/Table/Table.tsx [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    >
+      <table
+        ref={ref}
+        style={{ ...tableInlineVars }}
+        className={`${tableDefaults} ${
+          striped && "striped"
+        } ${overrideTableClass}`}
```
Comment (pruthvi2103) [inline lib/src/components/Table/Table.tsx:54]: Same as above, add the prop className

File: lib/src/components/Table/Table.tsx [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  });
+
+  return (
+    <div
+      style={{ ...tableContainerInlineVars }}
+      className={`${tableContainerDefaults} ${overrideTableContainerClass}`}
+    >
+      <table
+        ref={ref}
+        style={{ ...tableInlineVars }}
```
Comment (pruthvi2103) [inline lib/src/components/Table/Table.tsx:51]: Same as above, add the style

File: lib/src/components/Table/Td/Td.tsx [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+import { TdProps } from "./Td.types";
+
+const Td: ForwardRefRenderFunction<
+  HTMLTableCellElement,
+  TdProps
+> = ({ className, alignContent, children, ...nativeProps }, ref) => {
+  return (
+    <td
+      ref={ref}
+      style={{ textAlign: alignContent }}
```
Comment (pruthvi2103) [inline lib/src/components/Table/Td/Td.tsx:11]: same as above, style will get overridden

File: lib/src/components/Table/Tfoot/Tfoot.tsx [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  { children, backgroundColor = "#ddd", className, ...nativeProps },
+  ref
+) => {
+  return (
+    <tfoot
+      ref={ref}
+      style={assignInlineVars({
+        [tfootThemeVars.customBackgroundColor]: backgroundColor,
+      })}
+      className={`${footerDefaults} ${className}`}
```
Comment (pruthvi2103) [inline lib/src/components/Table/Tfoot/Tfoot.tsx:16]:
```suggestion
      className={`${footerDefaults} ${className || ""}`}
```

File: lib/src/components/Table/Thead/Thead.css.ts [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+
+export const [, theadThemeVars]: ITheadThemeVars = createTheme({
+  customBackgroundColor: "#ddd",
+});
+
+export const headerDefaults = style({});
+
+globalStyle(`${headerDefaults} tr th`, {
+  backgroundColor: theadThemeVars.customBackgroundColor,
+});
```
Comment (pruthvi2103) [inline lib/src/components/Table/Thead/Thead.css.ts:12]: add in seperate file

File: lib/src/components/Table/Thead/Thead.tsx [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+const Thead: ForwardRefRenderFunction<HTMLTableSectionElement, TheadProps> = (
+  { children, backgroundColor = "#ddd", className, ...nativeProps },
+  ref
+) => {
+  return (
+    <thead
+      ref={ref}
+      style={assignInlineVars({
+        [theadThemeVars.customBackgroundColor]: backgroundColor,
+      })}
```
Comment (pruthvi2103) [inline lib/src/components/Table/Thead/Thead.tsx:15]: merge the style from prop, it wil get overwritten

File: lib/src/components/Table/Thead/Thead.tsx [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+      className={`${headerDefaults} ${className}`}
```
Comment (pruthvi2103) [inline lib/src/components/Table/Thead/Thead.tsx:16]:
```suggestion
      className={`${headerDefaults} ${className || ""}`}
```

**Review by pruthvi2103 — CHANGES_REQUESTED** [review]
Added Comments

File: lib/src/components/Table/Table.tsx (hunk 3 above)
Comment (hafsahayath) [reply lib/src/components/Table/Table.tsx]: border-radius to the table doesn't apply if border-collapse is set to collapse, so I wrapped it around a div to achieve this.

Comment (pruthvi2103) [reply lib/src/components/Table/Table.tsx]: Oh ok got it, thanks!

**Review by pruthvi2103 — APPROVED** [review]
LGTM! Awesome work!!

**Comment by pruthvi2103** [comment]
I've approved the PR. Do you want to create a separate PR for documentation? Or will you work on this?

**Comment by hafsahayath** [comment]
> I've approved the PR. Do you want to create a separate PR for documentation? Or will you work on this?

I will create a separate PR for documentation, please merge this

**Pull request merged**